include/wil/com.h

```diff
@@ -3325,6 +3325,79 @@ WI_NODISCARD auto make_range(IEnumXxx* enumPtr)
 #endif // WIL_HAS_CXX_17
 #endif // WIL_ENABLE_EXCEPTIONS
 
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | WINAPI_PARTITION_SYSTEM)
+#ifdef __WIL_WIN32_HELPERS_INCLUDED
+
+/** RAII support for making cross-apartment (or cross process) COM calls with a timeout applied to them.
+ * When this is active any timed out calls will fail with an RPC error code such as RPC_E_CALL_CANCELED.
+ * This is a shared timeout that applies to all calls made on the current thread for the lifetime of
+ * the wil::rpc_timeout object.
+~~~
+{
+  auto timeout = wil::rpc_timeout(5000);
+  remote_object->BlockingCOMCall();
+  remote_object->AnotherBlockingCOMCall();
+}
+~~~
+Include wil/win32_helpers.h before wil/com.h to use this.
+*/
+class rpc_timeout

```

Comment:
>rpc_timeout [](http://example.com/codeflow?start=6&length=11)

I've always been skeptical of this name, given that this is absolutely COM-level cancelation, and could theoretically be used in inproc scenarios which don't involve the RPC runtime at all.

Reply:
Feel free to suggest a better name.  I don't mind renaming to something more accurate.

Reply:
The problem that I've been having is that `CoCancelCall` is failing with `RPC_E_NO_CONTEXT` which seems to mean that this truly does require RPC and does not work inproc across apartments.

Reply:
If the primary test thread is changed from MTA to STA then the RPC no context error goes away and CoCancelCall returns S_OK.  That does not seem to cause the thread to unblock, though.

Reply:
> Feel free to suggest a better name

com_timeout?

> That does not seem to cause the thread to unblock, though.

Yeah, I was doing some testing of my own against that last night and saw the same thing. Cancelation state gets set on the call, but that doesn't result in the client thread being able to return unilaterally. If I allow the callee STA to pump messages after the cancelation, the caller does end up getting RPC_E_CALL_CANCELED without invoking the server method, but yeah it was still blocked on the server to return.

Broadly speaking, issuing a cancelation doesn't guarantee that the cancelation will be seen/respected within any specific timeframe, so letter-of-the-law this is fine. But this sort of gap definitely limits the usefulness of canceling inproc calls.

I'll also note that one can use CoSetCancelObject to allow custom cancelation even for calls that don't pass 7through the COM runtime (as another example of non-RPC cancelation).